verified_boot_app_spec.md

```diff
@@ -0,0 +1,416 @@
+`ByProtocol` for `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL`, identifying all connected
+ESP-capable volumes. Each volume is queried via `OpenVolume()` to explore file
+systems for Secure Boot material and bootloaders.
+
+### 3.2 Key Discovery on Filesystems and Default Variables
+
+In this phase, the application scans all discovered ESPs for files that may
+contain usable public keys and certificates. It searches recursively for
+files with known cryptographic formats such as `.cer`, `.der`, or `.key`,
+using simple heuristics and file extension matching.
+
+Each candidate file is parsed using an internal X.509 ASN.1 parser. For
+every successfully decoded certificate or public key, the application
+extracts metadata such as:
+
+* Subject Common Name (CN) and Organization (O)
+* Issuer CN and O
+* Validity period
+* Key type and size
+* Fingerprint
+
+These keys are collected into a staging list of potential trust
+anchors.
+
+### 3.3 Bootloader Identification and Verification
+
+The application will parse the images of the bootloaders to obtain information
+about the signatures. The application will additionally validate bootloaders
+using `EFI_BOOT_SERVICES.LoadImage()`. Bootloaders that pass signature checks
+using are listed for potential trust decisions. For images with unknown but
+verifiable keys, the application initiates a trust prompt as described in the
+next section.
+
+The information about the bootloader should include:
+
+* The EFI image device path
+* Status of image verification
+* Signature list used during verification attempts
+* Associated certificates/keys
+
+### 3.4 Trust Decision and Key Enrollment
+
+For each unknown signing key discovered during verification, the application
+prompts the user with a detailed summary, including:
+
+* Key path
+* Certificate subject and issuer fields (if available)
+* Key fingerprint (SHA-256)
+* Origin (ESP, firmware default, audit log)
+
+With explicit user consent, the application enrolls the key into the `db`
+database using `SetVariable()`. Since the application operates in either
+Audit Mode or Setup Mode, authenticated write access is not required.
+
+Only the `db` (permitted image signers) may be updated during this phase.
+The `KEK` is not modified unless the user explicitly chooses to enroll
+Microsoft's default keys. No changes to `PK` are performed until the end
+of the provisioning workflow. This restriction ensures that the platform's
+trust anchors and ownership state are not altered prematurely.
+
+At the end of the provisioning workflow, if no Platform Key (`PK`) is
+present, the user is prompted to select a key for initial ownership. This
+key may be one of the previously trusted certificates or a new key
+supplied specifically for the purpose of establishing platform ownership.
+The application then sets the `PK` using authenticated variable services and
+enables Secure Boot by setting `SetupMode = 0` and `SecureBootEnable = 1`, as
+specified in the UEFI Specification.
```

Comment:
I think this contradicts with the app flow specified by app_flow.png - isn't the PK ephemeral? Or is this initial ownership key not the PK?

Reply:
Yes, I haven't completely cleaned up the text to match diagram yet